protostar/self/protostar_compatibility_with_project_checker.py

```diff
@@ -0,0 +1,56 @@
+from enum import Enum, auto
+from typing import Optional, Protocol
+
+from packaging import version
+
+ProtostarVersion = version.Version
+
+
+class DeclaredProtostarVersionProviderProtocol(Protocol):
+    def get_declared_protostar_version(self) -> Optional[ProtostarVersion]:
+        ...
```

Comment:
It would move the logic to the `composition_root.py`, and that's not the purpose of this file. Consider running `protostar --help`. Why should we read the protostar versions?

Reply:
If I shouldn't think about the composition root, then the instantiation of this class would have to be done in `ProtostarCLI` and the `ProtostarCLI` would have to coordinate reading versions, instantiating and passing versions to the instance of this class. It would couple `ProtostarCLI` with the protostar version readers and the checker. It's a functional programming mindset.

This is basically another OOP vs FP discussion.